rust/processor/src/bq_analytics/generic_parquet_processor.rs

```diff
@@ -153,84 +161,106 @@ where
-                    Err(anyhow!("Failed to upload file to GCS: {}", e))
-                },
-            };
+            if self.buffer_size_bytes >= self.max_buffer_size {
+                info!("Max buffer size reached, uploading to GCS.");
+                if let Err(e) = self.upload_buffer(gcs_client).await {
+                    error!("Failed to upload buffer: {}", e);
+                    return Err(e);
+                }
+                self.last_upload_time = Instant::now();
+            }
+
+            if self.last_upload_time.elapsed() >= self.upload_interval {
+                info!(
+                    "Time has elapsed more than {} since last upload.",
+                    self.upload_interval.as_secs()
+                );
+                if let Err(e) = self.upload_buffer(gcs_client).await {
+                    error!("Failed to upload buffer: {}", e);
+                    return Err(e);
+                }
+                self.last_upload_time = Instant::now();
+            }
         }
 
         PARQUET_HANDLER_BUFFER_SIZE
             .with_label_values(&[ParquetType::TABLE_NAME])
             .set(self.buffer.len() as i64);
         Ok(())
     }
+
+    pub async fn upload_buffer(&mut self, gcs_client: &GCSClient) -> Result<()> {
+        if self.buffer.is_empty() {
+            return Ok(());
+        }
+        let start_version = self.buffer.first().unwrap().version();
+        let last = self.buffer.last().unwrap();
+        let end_version = last.version();
+        let last_transaction_timestamp = naive_datetime_to_timestamp(last.get_timestamp());
+
+        let txn_version_to_struct_count =
+            process_struct_count_map(&self.buffer, &mut self.transaction_version_to_struct_count);
+
+        let new_file_path: PathBuf = PathBuf::from(format!(
+            "{}_{}.parquet",
+            ParquetType::TABLE_NAME,
+            Uuid::new_v4()
+        ));
+        rename(&self.file_path, &new_file_path)?; // this fixes an issue with concurrent file access issues
+
+        let struct_buffer = std::mem::take(&mut self.buffer);
+
+        let mut row_group_writer = self.writer.next_row_group()?;
+        struct_buffer
+            .as_slice()
+            .write_to_row_group(&mut row_group_writer)
+            .unwrap();
+        row_group_writer.close()?;
+        self.close_writer()?;
+
+        debug!(
+            table_name = ParquetType::TABLE_NAME,
+            start_version = start_version,
+            end_version = end_version,
+            "Max buffer size reached, uploading to GCS."
+        );
+        let bucket_root = PathBuf::from(&self.bucket_root);
+        let upload_result = upload_parquet_to_gcs(
+            gcs_client,
+            &new_file_path,
+            ParquetType::TABLE_NAME,
+            &self.bucket_name,
+            &bucket_root,
+        )
+        .await;
+        self.buffer_size_bytes = 0;
+        remove_file(&new_file_path)?;
```

Comment:
question: if gcs operation fails, how to recover? this reads to me that we panic, right? removal is to make sure we don't have any thing that's not cleaned up.

Reply:
if gcs operation fails after retries, and it doesn't panic her right? it cleans up and it bubbles up the error and panic at the `create_parquet_handler_loop`. and if it fails, we should restart from where it failed.